rp2040-hal/src/clocks/mod.rs

```diff
@@ -301,6 +305,46 @@ impl ClocksManager {
             .configure_clock(&self.system_clock, self.system_clock.freq())
     }
 
+    /// Approximates the frequency of the given clock source.
+    pub fn approximate_frequency<C: ClockSource>(&mut self, _trg_clk: C) -> HertzU32 {
+        // Wait for the frequency counter to be ready
+        while self.clocks.fc0_status.read().running().bit_is_set() {
+            core::hint::spin_loop()
+        }
+
+        // Set the speed of the reference clock in kHz.
+        self.clocks.fc0_ref_khz.write(|w| unsafe {
+            w.fc0_ref_khz()
+                .bits(self.reference_clock.get_freq().to_kHz())
+        });
+
+        // Corresponds to a 1ms test time, which seems to give good enough accuracy
+        self.clocks
+            .fc0_interval
+            .write(|w| unsafe { w.fc0_interval().bits(10) });
+
+        // We don't really care about the min/max, so these are just set to min/max values.
+        self.clocks
+            .fc0_min_khz
+            .write(|w| unsafe { w.fc0_min_khz().bits(0) });
+        self.clocks
+            .fc0_max_khz
+            .write(|w| unsafe { w.fc0_max_khz().bits(0xffffffff) });
+
+        // To measure rosc directly we use the value 0x03.
+        self.clocks
+            .fc0_src
+            .write(|w| w.fc0_src().variant(C::FCOUNTER_SRC));
+
+        // Wait until the measurement is ready
+        while self.clocks.fc0_status.read().done().bit_is_clear() {
+            core::hint::spin_loop()
+        }
+
+        let speed_hz = self.clocks.fc0_result.read().khz().bits() * 1000;
+        speed_hz.Hz()
```

Comment:
There are a few error bits in FC0_STATUS. This method could check those bits and return an error if any are set.

Reply:
Indeed, however it is unclear to me which is set and when.
- I'd assume `WAITING`, `RUNNING` and `DONE` are part of a onehot state machine.
  But is `DIED` part of it or is it only set when `DONE` is reached?
- I'd expect `FAIL` and `PASS` to be opposite but why would there be two flags for the same thing?
  * Can they ever be both set?

Reply:
There's some explanation in section 2.15.6.2. Using the frequency counter:

"The frequency counter can also be used in a test mode. This allows the hardware to check if the frequency is within
a minimum frequency and a maximum frequency, set in FC0_MIN_KHZ and FC0_MAX_KHZ. In this mode, the PASS bit in FC0_STATUS will be set when DONE is set if the frequency is within the specified range. Otherwise, either the FAST or SLOW bit will be set.
If the programmer attempts to count a stopped clock, or the clock stops running then the DIED bit will be set. If any of DIED, FAST, or SLOW are set then FAIL will be set."

In this case, FAST and SLOW don't apply. So the only realistic error case would be DIED. So a simple `Result<HertzU32, ()>` would be sufficient to communicate the possible return values.

Or we could define an `enum FrequencyCounterError { Died, Fast, Slow }` so we match the hardware, even if in practice two of the cases won't occur unless we also make the lower/upper frequency limits configurable.

Reply:
Strange: I tried to provoke a DIED response:
```
     let m1: Rate<u32, 1, 1> = clocks.measure_frequency(_trg_clk: &clocks.adc_clock, accuracy: FCAccuracy::_1kHz ); 
     info!("measurement 1: {} kHz", m1.to_kHz());                             
     clocks.adc_clock.disable(); 
     let m2: Rate<u32, 1, 1> = clocks.measure_frequency(_trg_clk: &clocks.adc_clock, accuracy: FCAccuracy::_1kHz ); 
     info!("measurement 2: {} kHz", m2.to_kHz()); 
```
But I got a successful measurement:
```
└─ rp2040_project_template::__cortex_m_rt_main @ src/main.rs:27
INFO  measurement 1: 48000 kHz
└─ rp2040_project_template::__cortex_m_rt_main @ src/main.rs:59
INFO  measurement 2: 0 kHz
```

Which isn't exactly wrong - a disabled clock is a 0kHz clock. But then, when does DIED get set?

Reply:
I think it does when it sees a couple clock cycles and then it stops while measuring is taking place.